kernel/src/grant.rs

```diff
@@ -243,11 +252,19 @@ impl<T: Default> Grant<T> {
     where
         F: Fn(&mut Owned<T>),
     {
-        self.kernel.process_each(|process| unsafe {
-            if let Some(grant_ptr_ref) = process.grant_ptr(self.grant_num) {
-                let root_ptr = *(grant_ptr_ref as *mut *mut T);
-                if !root_ptr.is_null() {
-                    let mut root = Owned::new(root_ptr, process.appid());
+        self.kernel.process_each(|process| {
+            if let Some(grant_ptr) = process.grant_ptr(self.grant_num) {
+                // Check if this grant has been allocated yet
+                let grant = {
+                    if grant_ptr.is_null() {
+                        None
+                    } else {
+                        // Requirement: If grant has been allocated, it is sized for T
+                        unsafe { Some(NonNull::new_unchecked(*grant_ptr as *mut T)) }
+                    }
```

Comment:
Same thing as above. This can be simplified as:

```rust
NonNull::new(grant_ptr).cast::<T>()
```

and then probably calling `.map()` on this option rather than using `if let Some`.